flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4412,30 +4511,64 @@ class ArrayExprLowering {
-            // the explicit loop control variables.
-            for (const auto &ss : subs)
+            // scalar expression it is either loop-invariant or a function
+            // of the explicit loop control variables.
+            for (const auto &ss : subs) {
               if (auto *intExpr = std::get_if<
-                      Fortran::evaluate::IndirectSubscriptIntegerExpr>(&ss.u)) {
-                if (intExpr->value().Rank() > 0) {
-                  explicitImpliedShape.push_back(feShape[ii++]);
+                      Fortran::evaluate::IndirectSubscriptIntegerExpr>(&ss.u))
+                if (intExpr->value().Rank() > 0)
                   vectorCoor[vi++] = genarr(intExpr->value());
-                }
-              } else {
-                // This is a triple which may be using an explicit control
-                // variable.
-                explicitImpliedShape.push_back(feShape[ii++]);
-              }
+            }
+            explicitImpliedLoopBounds = [=]() {
+              llvm::SmallVector<mlir::Value> result;
+              unsigned ii = 0;
+              for (const auto &ss : subs)
+                std::visit(
+                    Fortran::common::visitors{
+                        [&](const Fortran::evaluate::
+                                IndirectSubscriptIntegerExpr &intExpr) {
+                          if (intExpr.value().Rank() > 0)
+                            result.push_back(builder.createConvert(
+                                loc, idxTy, desShape[ii++]));
+                        },
+                        [&](const Fortran::evaluate::Triplet &triple) {
+                          // This is a triple which may be using an
+                          // explicit control variable.
+                          auto ou = triple.upper();
+                          auto up = builder.createConvert(
+                              loc, idxTy,
+                              ou.has_value() ? fir::getBase(asScalar(*ou))
+                                             : desShape[ii]);
+                          auto ol = triple.lower();
+                          auto lo =
+                              ol.has_value()
+                                  ? builder.createConvert(
+                                        loc, idxTy, fir::getBase(asScalar(*ol)))
+                                  : builder.createIntegerConstant(loc, idxTy,
+                                                                  1);
+                          auto step = builder.createConvert(
+                              loc, idxTy,
+                              fir::getBase(asScalar(triple.stride())));
+                          auto diff = builder.create<mlir::SubIOp>(loc, up, lo);
+                          auto sum =
+                              builder.create<mlir::AddIOp>(loc, diff, step);
+                          mlir::Value count =
+                              builder.create<mlir::SignedDivIOp>(loc, sum,
```

Comment:
This triplet extent computation pattern will definitely deserve to be centralized in a helper at some point. FYI, I think it does not cover the legal case of `i:j:k` where `j-i` and k sign differ, in which case the computed extent must be zero.

Reply:
Agreed. I'll put that on my list of TODOs.